src/intptrcast.rs

```diff
@@ -72,12 +74,45 @@ impl<'mir, 'tcx> GlobalStateInner {
                     None
                 }
             }
-        };
-        // Pointers created from integers are untagged.
-        Pointer::new(
-            alloc_id.map(|alloc_id| Tag { alloc_id, sb: SbTag::Untagged }),
-            Size::from_bytes(addr),
-        )
+        }?;
+
+        match global_state.provenance_mode {
+            ProvenanceMode::Legacy => Some(alloc_id),
+            ProvenanceMode::Permissive if global_state.exposed.contains(&alloc_id) =>
+                Some(alloc_id),
+            _ => None,
```

Comment:
Always returning `None` for strict provenance after doing all the work to compute the ID seems kind of a waste. We probably should handle that in the caller though.
```suggestion
        // In legacy mode, we consider all allocations exposed.
        if global_state.provenance_mode == ProvenanceMode::Legacy || global_state.exposed.contains(&alloc_id) {
            Some(alloc_id)
        } else {
            None
        }
```